[ui-side compositing] Add check for layer after scaling CATransaction completed

nmoucht opened the pull request:
#### e2f320571c31401849a94f07ce06765dc4dd1e7c
<pre>
[ui-side compositing] Add check for layer after scaling CATransaction completed
<a href="https://bugs.webkit.org/show_bug.cgi?id=255816">https://bugs.webkit.org/show_bug.cgi?id=255816</a>
rdar://108234128

Reviewed by Simon Fraser.

Add check for layer after scaling CATransaction completed. Also hold a
RefPtr to the WebPageProxy to ensure that the drawing area still exists
after the completion block finishes.

* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::commitTransientZoom):

Canonical link: <a href="https://commits.webkit.org/263643@main">https://commits.webkit.org/263643@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39e43069f277473e176569729e4dc70e7f967e3e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5314 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5450 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5638 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6849 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5356 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5672 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5431 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5948 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5413 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5486 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4763 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6875 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2968 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/11780 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4832 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4844 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6461 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5269 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/4329 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4736 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8831 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/598 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5096 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
```diff
@@ -305,7 +305,10 @@ explicit RemoteLayerTreeDisplayLinkClient(WebPageProxyIdentifier pageID)
+    [CATransaction setCompletionBlock:[this, scale, constrainedOrigin, transform] () {
+        CALayer *layerForPageScale = remoteLayerTreeHost().layerForID(m_pageScalingLayerID);
+        if (!layerForPageScale)
+            return;
```
We should still set `m_transactionIDAfterEndingTransientZoom` and call `m_webPageProxy.send(Messages::DrawingArea::CommitTransientZoom...` right?

nmoucht replied:
Hmm im not sure exactly what would cause this layer to go away. Would it still make sense to commit the zoom in the web process in this case?

smfr replied:
This may happen on page navigation. Leaving the web process without a committed zoom sounds like a bad idea.

nmoucht replied:
Ah ok gotcha. Updated.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
```diff
@@ -304,13 +304,19 @@ explicit RemoteLayerTreeDisplayLinkClient(WebPageProxyIdentifier pageID)
-        [layerForPageScale removeAnimationForKey:@"transientZoomCommit"];
-        m_transactionIDAfterEndingTransientZoom = nextLayerTreeTransactionID();
-        m_webPageProxy.send(Messages::DrawingArea::CommitTransientZoom(scale, constrainedOrigin), m_identifier);
+    [CATransaction setCompletionBlock:[scale, constrainedOrigin, transform, protectedWebPageProxy] () {
```
You can write this as: `[..., protectedWebPageProxy = Ref { m_webPageProxy }]`

nmoucht replied:
Fixed.

nmoucht committed:
[ui-side compositing] Add check for layer after scaling CATransaction completed
https://bugs.webkit.org/show_bug.cgi?id=255816
rdar://108234128

Reviewed by Simon Fraser.

Add check for layer after scaling CATransaction completed. Also hold a
RefPtr to the WebPageProxy to ensure that the drawing area still exists
after the completion block finishes.

* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::commitTransientZoom):

Canonical link: https://commits.webkit.org/263643@main
